File: cpp/2506_similar_pairs/similar_pairs.cpp
#include <cassert>
#include <string>
#include <tuple>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
 public:
  int similarPairs(vector<string>& words) {
    int res = 0;
    unordered_map<int, int> cnt;
    for (const string& word : words) {
      int state = 0;
      for (char c : word) {
        state |= 1 << (c - 'a');
      }
      res += cnt[state];
      cnt[state]++;
    }
    return res;
  }
};

int main() {
  vector<tuple<vector<string>, int>> tests{
      {{"aba", "aabb", "abcd", "bac", "aabc"}, 2},
      {{"aabb", "ab", "ba"}, 3},
      {{"nba", "cba", "dba"}, 0},
  };
  
  for (auto &[words, ans] : tests) {
    assert(Solution().similarPairs(words) == ans);
  }
}

File: golang/2506_similar_pairs/similar_pairs.go
// Package main ...
package main

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func similarPairs(words []string) int {
	res := 0
	cnt := make(map[int]int)
	for _, word := range words {
		state := 0
		for _, c := range word {
			state |= 1 << (c - 'a')
		}
		res += cnt[state]
		cnt[state]++
	}
	return res
}

func main() {
	tests := []struct {
		words []string
		ans   int
	}{
		{[]string{"aba", "aabb", "abcd", "bac", "aabc"}, 2},
		{[]string{"aabb", "ab", "ba"}, 3},
		{[]string{"nba", "cba", "dba"}, 0},
	}

	for index, test := range tests {
		assert.Equal(&testing.T{}, test.ans, similarPairs(test.words), index)
	}
}
